rust/libnewsboat/src/fslock.rs

```diff
@@ -0,0 +1,98 @@
+use crate::logger::{self, Level};
+use std::fs::{self, File, OpenOptions};
+use std::io::{Read, Write};
+use std::os::unix::fs::OpenOptionsExt;
+use std::os::unix::io::AsRawFd;
+use std::path::{Path, PathBuf};
+use std::process;
+
+fn remove_lock(lock_path: &Path) {
+    fs::remove_file(lock_path).ok();
+    log!(
+        Level::Debug,
+        "FsLock: removed lockfile {}",
+        lock_path.display()
+    );
+}
+
+#[derive(Default)]
+pub struct FsLock {
+    lock_path: PathBuf,
+    lock_file: Option<File>,
+}
+
+impl Drop for FsLock {
+    fn drop(&mut self) {
+        if self.lock_file.is_some() {
+            remove_lock(&self.lock_path);
+        }
+    }
+}
+
+impl FsLock {
+    pub fn try_lock(&mut self, new_lock_path: &Path, pid: &mut libc::pid_t) -> bool {
+        if self.lock_file.is_some() && self.lock_path == new_lock_path {
+            return true;
+        }
+
+        // pid == 0 indicates that something went majorly wrong during locking
+        *pid = 0;
+
+        log!(
+            Level::Debug,
+            "FsLock: trying to lock `{}'",
+            new_lock_path.display()
+        );
+
+        // first we open (and possibly create) the lock file
+        let mut options = OpenOptions::new();
+        options.mode(0o600).truncate(true);
```

Comment:
I don't think this will work the way the original code did, because this:

1. won't create a file if it's missing;
2. will immediately truncate the file, so if it's locked, FsLock won't be able to read the PID from it.

I believe you should add back `.create(true)`, and move truncation down into the `then` branch.

Reply:
So `options.create(true).read(true).write(true).mode(0o600).truncate(true)`?

Reply:
No, just `options.create(true).read(true).write(true).mode(0o600)`. Truncation has to happen *after* the `lockf()` call.

Reply:
Ah